server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/matchers/Or.java

```diff
@@ -37,18 +41,17 @@
 public class Or extends GenericCompositeMatcher {
 
     @Override
-    public Collection<MailAddress> match(Mail mail) throws MessagingException {
-        HashSet<MailAddress> result = new HashSet<>();
-        for (Matcher matcher : getMatchers()) {
-            Collection<MailAddress> matchResult = matcher.match(mail);
-            if (matchResult != null) {
-                result.addAll(matchResult);
-            }
-            if (result.size() == mail.getRecipients().size()) {
-                break;
-            }
-        }
-        return result;
+    public Collection<MailAddress> match(Mail mail) {
+        return getMatchers().stream()
+            .flatMap(Throwing.<Matcher, Stream<MailAddress>>function(matcher -> applyMatcher(matcher, mail))
+                .sneakyThrow())
+            .collect(Collectors.toSet());
```

Comment:
Why not an immutableSet?

Reply:
Declaring something immutable should, in my mind, express an intent. Here, I won't keep any reference to this list and I don't care if the caller want to modify it. I also think that it's a conservative choice to not change from mutable to immutable because it would be a semantic change (think about a caller actually relying on the fact it's mutable, it will just break the existing code).
We already had that conversation in the past and we decided that consistent was more important then fine-grain decision regarding collectors and that we should always collect to immutable datastructures.
However, in the case of a refactoring, where we should never change the semantic, I think it make sense to collect to a mutable set here.
WDYT?